[TEST] Enforce skip headers when needed

javanna opened the pull request:
The java yaml test runner supports sending request headers, yet not all clients support headers. This commit makes sure that we enforce adding a skip section with feature "headers" whenever headers are used in a do section as part of a test. That decreases the chance for new tests to break client builds due to the missing skip section.

Closes #34650

javanna committed:
[TEST] Enforce skip headers when needed

The java yaml test runner supports sending request headers, yet not all clients support headers. This commit makes sure that we enforce adding a skip section with feature "headers" whenever headers are used in a do section as part of a test. That decreases the chance for new tests to break client builds due to the missing skip section.

Closes #34650

elasticmachine commented:
Pinging @elastic/es-core-infra

atorok commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java:
```diff
@@ -98,6 +98,12 @@ public void addExecutableSection(ExecutableSection executableSection) {
+            if (false == doSection.getApiCallSection().getHeaders().isEmpty()
+                    && false == skipSection.getFeatures().contains("headers")) {
+                throw new IllegalArgumentException("Attempted to add a [do] with a [headers] section without a corresponding "
+                    + "[skip] so runners that do not support the [headers] section can skip the test at line ["
```
Maybe worth mentioning the feature has the same name to save people some digging ?
```suggestion
                    + "[skip: {"feature": "headers"}] so runners that do not support the [headers] section can skip the test at line ["
```

alpar-t reviewed (approved):
LGTM just a suggestion.

javanna committed:
adjusted tests

javanna committed:
remove constructor

javanna committed:
remove connection manager public constructor

javanna commented:
retest this please

javanna committed:
Merge branch 'master' into test/yaml_enforce_skip_headers

javanna committed:
Revert "remove connection manager public constructor"

This reverts commit 0ff7065c82cacc99e3c661b22ada5814aecdb4ce.

javanna committed:
Revert "remove constructor"

This reverts commit e681ef548b5cc269054d16c09acb9b3044387be8.

javanna committed:
Revert "adjusted tests"

This reverts commit 5ac601063e0539a6615b461b2f475af076bb6219.

javanna committed:
address comment

javanna committed:
increase existings tests coverage

javanna committed:
fix broken test

javanna committed:
fix test

javanna committed:
fix test

javanna committed:
Merge branch 'master' into test/yaml_enforce_skip_headers

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSection.java:
```diff
@@ -88,14 +88,21 @@ public void addExecutableSection(ExecutableSection executableSection) {
+            }
+            if (false == doSection.getApiCallSection().getHeaders().isEmpty()
+                    && false == skipSection.getFeatures().contains("headers")) {
+                throw new IllegalArgumentException("Attempted to add a [do] with a [headers] section without a corresponding "
```
:+1:

javanna committed:
adjusted tests

javanna committed:
Merge branch 'master' into test/yaml_enforce_skip_headers

javanna committed:
fix test

javanna committed:
improve validation of do sections

javanna committed:
Revert "adjusted tests"

This reverts commit f8ecf2ec820d54285f8b4b3c05caa46b2e884674.

javanna committed:
Revert "fix test"

This reverts commit 7083d3e32f236242e7654f3bca21dab65fbbf0f6.

javanna committed:
fix bad merge

javanna committed:
Revert "fix test"

This reverts commit b4181a25dedf02ef006a12ffea8dcf26ef1419e7.

javanna committed:
reverted change

javanna committed:
remove empty line

javanna commented:
@alpar or @nik9000 can you maybe have another look? I had to move the validation code higher-up, as this change made another bug surface: we don't look at the skip section from setup or teardown in our validation, and we have quite some skip headers in our setup sections, so false positives without adapting the validation code. Let me know what you think.

javanna committed:
checkstyle

javanna commented:
retest this please

javanna committed:
Merge branch 'master' into test/yaml_enforce_skip_headers

javanna committed:
fix more tests and improve validation

javanna committed:
fix test

javanna committed:
fix test

javanna committed:
fix ml tests

javanna committed:
make setup, teardown and test sections immutable and improve validation

We now have more accurate validation of setup, teardown and test sections as for missing skip feature sections

javanna committed:
fix tests

javanna committed:
fix more tests

javanna committed:
Merge branch 'master' into test/yaml_enforce_skip_headers

javanna committed:
Merge branch 'master' into test/yaml_enforce_skip_headers

javanna committed:
fix bad merge

javanna committed:
improve test
